plugins/factory-plugin/src/git.ts

```diff
@@ -0,0 +1,133 @@
+        const grep = spawn('grep', ['\*']);
+
+        gitBranchCmd.stdout.pipe(grep.stdin);
+
+        let output = '';
+        grep.stdout.on('data', function(data: string) {
+            output += data.toString();
+        });
+        grep.on('close', async function(code: any) {
+            output = output.trim();
+            const gitBranch: GitBranch | undefined = parseGitBranch(output);
+            if (gitBranch && gitBranch.upstreamBranch) {
+                gitBranch.upstreamBranch.remoteURL = await getRemoteURL(gitBranch.upstreamBranch.remote, projectPath);
+            }
+
+            resolve(gitBranch);
+        });
+        grep.on('error', function(err: any) { reject(err); });
+    });
+}
+
+export function getRemoteURL(remote: string, projectPath: string): Promise<string | undefined> {
+    return new Promise<string | undefined>((resolve, reject) => {
+
+        const gitConfigUrl = spawn('git', ['config', '--get', `remote.${remote}.url`], { cwd: projectPath });
+        let result = '';
+        gitConfigUrl.stdout.on('data', function(data: string) {
+            result += data.toString();
+        });
+        gitConfigUrl.on('close', function(code: any) {
+            result = result.trim();
+            resolve(result);
+        });
+        gitConfigUrl.on('error', function(err: any) { reject(err); });
+    });
+}
+
+export function parseGitBranch(gitBranchvvOutput: string): GitBranch | undefined {
+
+    const star = '[\*]?';
+    const spaces = '[\\s]+';
+    const nonWhiteSpaces = '[^\\s]+';
+    const localBranch = nonWhiteSpaces;
+    const ref = nonWhiteSpaces;
+    const branchOrRemote = '[^\\s^/]+';
+
+    const regexp = new RegExp(
+        `^(${star})${spaces}(${localBranch})${spaces}(${ref})${spaces}` +
+        `(\\[(${branchOrRemote})\\/(${branchOrRemote})(: behind ([0-9]+))?(: ahead ([0-9]+))?\\])?.*`);
+
+    // for instance:
+    // *  master     e619393 [origin/master: behind 4] call 'onTimeout' callback when notification closed on timeout
+    // * vplugin-id 57f328a [sunix/vplugin-id] Moving openfile command args to the right command."
+
+    const result: RegExpMatchArray | null = gitBranchvvOutput.match(regexp);
+
+    if (!result) {
+        return undefined;
+    }
+
+    let aheadBehind = undefined;
+
+    if (result[8]) {
+        aheadBehind = { behind: parseInt(result[8], 10) };
+    }
+    if (result[10]) {
+        aheadBehind = { ahead: parseInt(result[10], 10) };
+    }
+
+    let upstreamBranch = undefined;
+    if (result[4]) {
+        upstreamBranch = {
+            remote: result[5], branch: result[6]
+        };
+    }
+    return {
+        isHEAD: (result[1] === '*'),
+        localBranch: result[2],
+        upstreamBranch: upstreamBranch,
+        aheadBehind: aheadBehind,
+        ref: result[3]
+    };
+
+}
+
+export function getGitRootFolder(uri: string): string {
+    if (uri.endsWith('/.git/config')) {
```

Comment:
uri.lastIndexOf('.git/') / uri.substring(0, uri.lastIndexOf('.git/'))) ?

or regexp or usage of ```git rev-parse --show-toplevel``` 

(so if you need to track other files in .git folder, no need to add new lines)

Reply:
problem with `git rev-parse --show-toplevel` is that it seems it does not work under .git folder :(